core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -282,14 +282,18 @@ public void commit() {
           .run(taskOps -> {
             Snapshot newSnapshot = apply();
             newSnapshotId.set(newSnapshot.snapshotId());
-            TableMetadata updated;
-            if (stageOnly) {
-              updated = base.addStagedSnapshot(newSnapshot);
+            TableMetadata.Builder update = TableMetadata.buildFrom(base);
+            if (base.snapshot(newSnapshot.snapshotId()) != null) {
+              // this is a rollback operation
+              update.setBranchSnapshot(newSnapshot.snapshotId(), SnapshotRef.MAIN_BRANCH);
```

Comment:
@wypoon and @sririshindra, I think this fixes the problem from #4088. The previous call to `replaceCurrentSnapshot` always used `setCurrentSnapshot(Snapshot, String)` rather than `setCurrentSnapshot(long, String)`. This detects when the snapshot already exists and calls the correct method.

Reply:
Yes, that makes sense.
I cherry-picked some prerequisite commits, and then applied this change to our 0.13.0 branch, and reran @sririshindra's rollback test, and it passed.